core/dnsserver/address.go

```diff
@@ -94,3 +94,82 @@ const (
 	TransportTLS  = "tls"
 	TransportGRPC = "grpc"
 )
+
+const keyPartsSeparator = "##"
+
+type addrKey struct {
+	Transport string // dns, tls or grpc
+	Zone      string
+	Address   string
+	Port      string
+}
+
+func (k *addrKey) asKey() string {
+	return k.Transport + keyPartsSeparator + k.Zone + keyPartsSeparator + k.Address + keyPartsSeparator + k.Port
+}
+
+func parseAddrKey(key string) (*addrKey, error) {
+	keyParts := strings.Split(key, keyPartsSeparator)
+	if len(keyParts) != 4 {
+		return nil, fmt.Errorf("provided value is not a key for addrKey, expect 4 parts joined by '%s' : %s", key, keyPartsSeparator)
+	}
+	return &addrKey{keyParts[0], keyParts[1], keyParts[2], keyParts[3]}, nil
+}
+
+func (k *addrKey) isUnbound() bool {
+	return k.Address == ""
+}
+
+func (k *addrKey) copyAsUnbound() *addrKey {
+	return &addrKey{Zone: k.Zone, Address: "", Port: k.Port, Transport: k.Transport}
+
+}
+
+// String return the string representation of this ZoneAddr - for print
+func (k *addrKey) String() string {
+	if k.Address == "" {
+		return k.Transport + "://" + k.Zone + ":" + k.Port
+	}
+	return k.Transport + "://" + k.Zone + ":" + k.Address + ":" + k.Port
+}
+
+// Build a Validator that rise error if the bound addresses for listeners are overlapping
+// we consider that an unbound address is overlapping all bound addresses for same zome, zame port
+type zoneAddrOverlapValidator struct {
+	registeredAddr map[string]string // each zoneAddr is registered once by its key (bound and unbound)
+	unboundOverlap map[string]string // the unbound equiv ZoneAdddr is registered by its original key (bound or unbound)
+}
+
+func newZoneAddrOverlapValidator() *zoneAddrOverlapValidator {
+	return &zoneAddrOverlapValidator{registeredAddr: make(map[string]string), unboundOverlap: make(map[string]string)}
+}
+
+func (c *zoneAddrOverlapValidator) registerAndCheck(k *addrKey) (same bool, overlap bool, overlapKey string) {
+
+	// we consider there is an overlap if:
+	//  - the current zoneAddr is already registered is already registered for the same zone/port
+	//  - the current zoneAddr is unbound and a non unbound is already registered for the same zone/port
+	//  - the current zoneAddr is bound and an unbound is already registered for the same zone/port
+	key := k.asKey()
+	if _, ok := c.registeredAddr[key]; ok {
+		// exact same zone already registered
+		return true, false, ""
+	}
+	mkey := k.copyAsUnbound().asKey()
+	if already, ok := c.unboundOverlap[mkey]; ok {
```

Comment:
as the plugin is called `bind` this is actually `bindOverlap`?

Reply:
I will change the name.